Sources/Client/WebSocket/Provider/URLSessionWebSocketProvider.swift

```diff
@@ -0,0 +1,95 @@
+    
+    init(request: URLRequest, callbackQueue: DispatchQueue) {
+        self.request = request
+        self.callbackQueue = callbackQueue
+    }
+    
+    func connect() {
+        let session = URLSession(configuration: URLSessionConfiguration.default, delegate: self, delegateQueue: nil)
+        task = session.webSocketTask(with: request)
+        doRead()
+        task?.resume()
+    }
+    
+    func disconnect() {
+        isConnected = false
+        task?.cancel(with: .abnormalClosure, reason: nil)
+    }
+    
+    func sendPing() {
+        task?.sendPing { _ in }
+    }
+    
+    private func doRead() {
+        task?.receive { [weak self] result in
+            guard let self = self else {
+                return
+            }
+            
+            switch result {
+            case .success(let message):
+                if case .string(let string) = message {
+                    self.callDelegateInCallbackQueue { $1?.websocketDidReceiveMessage($0, message: string) }
+                }
+            case .failure(let error):
+                self.disconnect(with: WebSocketProviderError(reason: error.localizedDescription,
+                                                             code: (error as NSError).code,
+                                                             providerType: URLSessionWebSocketProvider.self,
+                                                             providerError: error))
+            }
+            
+            self.doRead()
+        }
+    }
+    
+    public func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
+        isConnected = true
+        callDelegateInCallbackQueue { $1?.websocketDidConnect($0) }
+    }
+    
+    func urlSession(_ session: URLSession,
+                    webSocketTask: URLSessionWebSocketTask,
+                    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
+                    reason: Data?) {
+        var error: WebSocketProviderError?
+        
+        if let reasonData = reason, let reasonString = String(data: reasonData, encoding: .utf8) {
+            error = WebSocketProviderError(reason: reasonString,
+                                           code: 0,
+                                           providerType: URLSessionWebSocketProvider.self,
+                                           providerError: nil)
+        }
+        
+        disconnect(with: error)
+    }
+    
+    private func disconnect(with error: WebSocketProviderError?) {
+        isConnected = true
+        callDelegateInCallbackQueue { $1?.websocketDidDisconnect($0, error: error) }
+    }
+    
+    private func callDelegateInCallbackQueue(execute block: @escaping (WebSocketProvider, WebSocketProviderDelegate?) -> Void) {
+        callbackQueue.async { [weak self] in
+            if let self = self {
+                block(self, self.delegate)
+            }
+        }
```

Comment:
Why `[weak self]` here?

Reply:
`self -> callback -> self`
If the hosted app will set user again or token will be refreshed the web socket provider will be deallocated.
It will be great if you tell me why it's no needed here.